core/src/main/scala/kafka/log/Log.scala

```diff
@@ -763,7 +764,7 @@ class Log(@volatile var dir: File,
         if (truncatedBytes > 0) {
           // we had an invalid message, delete all remaining log
           warn(s"Corruption found in segment ${segment.baseOffset}, truncating to offset ${segment.readNextOffset}")
-          unflushed.foreach(deleteSegment)
+          removeAndDeleteSegments(unflushedSegments)
```

Comment:
Hmm, this doesn't seem right. We want to only delete from the first segment having truncatedBytes > 0, not necessarily all unflushedSegments.

Reply:
Also, we need to short circuit the loop when we hit the first segment with truncatedBytes > 0.

Reply:
Thanks for catching this! I caught this earlier today and pushed a fix in 509b916. I will push another commit to make the short circuit more explicit.